### PR TITLE
Add type hints in guided_backprop_deconvnet and test files

### PR DESCRIPTION
**Pull request by cheetah2216**
Add type hints in guided_backprop_deconvnet and test files

Summary:
Add type hints in guided_backprop_deconvnet and test files
Test Plan:
./scripts/run_mypy.sh
No new errors
Existing error
captum/attr/_core/feature_permutation.py:31: error: Argument "forward_func" to "__init__" of "FeatureAblation" has incompatible type "Optional[Callable[..., Any]]"; expected "Callable[..., Any]"

python -m unittest -v tests.attr.test_deconvolution
ok
test_simple_input_conv_deconv (tests.attr.test_deconvolution.Test) ... ok
test_simple_input_conv_neuron_deconv (tests.attr.test_deconvolution.Test) ... ok
test_simple_multi_input_conv_deconv (tests.attr.test_deconvolution.Test) ... /home/subodhlele/github/captum/captum/attr/_utils/gradient.py:32: UserWarning: Input Tensor 1 did not already require gradients, required_grads has been set automatically.
  "required_grads has been set automatically." % index
ok
test_simple_multi_input_conv_neuron_deconv (tests.attr.test_deconvolution.Test) ... ok

----------------------------------------------------------------------
Ran 5 tests in 0.118s

OK

python -m unittest -v tests.attr.test_guided_backprop

Reviewers:

Subscribers:

Tasks:
[captum-typehint] Adding Type Hints to Captum: guided_backprop_deconvnet.py
Tags:

### CODE REVIEW

File: captum/attr/_core/guided_backprop_deconvnet.py [hunk 1]
```diff
@@ -2,29 +2,41 @@
...
         )
 
-    def attribute(self, inputs, target=None, additional_forward_args=None):
+    def attribute(
+            self,
+            inputs: TensorOrTupleOfTensors,
+            target: Optional[
+                Union[int, Tuple[int, ...], Tensor, List[Tuple[int, ...]]]
+            ] = None,
+            additional_forward_args: Any = None):
```
Comment (vivekmig) [inline captum/attr/_core/guided_backprop_deconvnet.py]: This should return TensorOrTupleOfTensors, and similarly for GuidedBackprop and Deconvolution

File: tests/attr/test_deconvolution.py [hunk 2]
```diff
@@ -10,8 +10,11 @@
...
 )
 
+from typing import Any, Tuple, Union, List
+from torch.nn import Module
 from .helpers.basic_models import BasicModel_ConvNet_One_Conv
 from .helpers.utils import assertTensorAlmostEqual, BaseTest
+from captum.attr._utils.typing import TensorOrTupleOfTensors
 
 
 class Test(BaseTest):
```
Comment (vivekmig) [inline tests/attr/test_deconvolution.py]: The test methods below need to be type hinted as well, most of them will just have return -> None.

File: tests/attr/test_deconvolution.py [hunk 3]
```diff
@@ -68,7 +71,12 @@ def test_deconv_matching(self):
         inp = 100.0 * torch.randn(1, 1, 4, 4)
         self._deconv_matching_assert(net, net.relu2, inp)
 
-    def _deconv_test_assert(self, model, test_input, expected, additional_input=None):
+    def _deconv_test_assert(
+            self,
+            model: Module,
+            test_input: TensorOrTupleOfTensors,
+            expected: List,
```
Comment (vivekmig) [inline tests/attr/test_deconvolution.py]: Can you make the expected hint more specific, e.g List[List[float]]? It will likely be a Union including Tuples. Same for the functions below.

File: tests/attr/test_guided_backprop.py [hunk 4]
```diff
@@ -1,102 +1,293 @@
 #!/usr/bin/env python3
+import warnings
```
Comment (vivekmig) [inline tests/attr/test_guided_backprop.py]: It looks like this file got overwritten, can you replace with the original file and add type hints to that, similar to the other test file? Thanks!

File: captum/attr/_core/guided_backprop_deconvnet.py (hunk 1 above)
Comment (cheetah2216) [reply captum/attr/_core/guided_backprop_deconvnet.py]: done

File: tests/attr/test_deconvolution.py (hunk 2 above)
Comment (cheetah2216) [reply tests/attr/test_deconvolution.py]: done

File: tests/attr/test_deconvolution.py (hunk 3 above)
Comment (cheetah2216) [reply tests/attr/test_deconvolution.py]: done

File: tests/attr/test_guided_backprop.py (hunk 4 above)
Comment (cheetah2216) [reply tests/attr/test_guided_backprop.py]: done

File: tests/attr/test_deconvolution.py [hunk 5]
```diff
@@ -63,12 +66,20 @@ def test_simple_multi_input_conv_neuron_deconv(self):
...
         inp = 100.0 * torch.randn(1, 1, 4, 4)
         self._deconv_matching_assert(net, net.relu2, inp)
 
-    def _deconv_test_assert(self, model, test_input, expected, additional_input=None):
+    def _deconv_test_assert(
+        self,
+        model: Module,
+        test_input: TensorOrTupleOfTensors,
+        expected: Union[
+            Tuple[List[List[float]]], Tuple[List[List[float]], List[List[float]]]
```
Comment (vivekmig) [inline tests/attr/test_deconvolution.py]: nit: This could be simpler by just having Tuple[List[List[float]], ...] for arbitrary length tuples.

File: tests/attr/test_deconvolution.py [hunk 6]
```diff
@@ -77,8 +88,16 @@ def _deconv_test_assert(self, model, test_input, expected, additional_input=None
...
     def _neuron_deconv_test_assert(
-        self, model, layer, neuron_index, test_input, expected, additional_input=None
-    ):
+        self,
+        model: Module,
+        layer: Module,
+        neuron_index: Union[int, Tuple[int, ...]],
+        test_input: TensorOrTupleOfTensors,
+        expected: Union[
+            Tuple[List[List[float]]], Tuple[List[List[float]], List[List[float]]]
```
Comment (vivekmig) [inline tests/attr/test_deconvolution.py]: Same here.

File: tests/attr/test_guided_backprop.py [hunk 7]
```diff
@@ -61,14 +64,20 @@ def test_simple_multi_input_conv_neuron_gb(self):
...
         self._guided_backprop_matching_assert(net, net.relu2, inp)
 
     def _guided_backprop_test_assert(
-        self, model, test_input, expected, additional_input=None
-    ):
+        self,
+        model: Module,
+        test_input: TensorOrTupleOfTensors,
+        expected: Union[
+            Tuple[List[List[float]]], Tuple[List[List[float]], List[List[float]]]
```
Comment (vivekmig) [inline tests/attr/test_guided_backprop.py]: Same here and below.

File: captum/attr/_core/guided_grad_cam.py [hunk 8]
```diff
@@ -197,7 +197,7 @@ def attribute(
             target=target,
             additional_forward_args=additional_forward_args,
         )
-        output_attr = []
+        output_attr: List[Any] = []
```
Comment (vivekmig) [inline captum/attr/_core/guided_grad_cam.py]: This would be more specifically: List[Optional[Tensor]]

**Comment by cheetah2216** [comment]
> Looks great, thanks! Just a couple of minor things, and it'll be ready to Import to Phabricator.

Thanks, updated
